### ISSUE TITLE
Should str.replace accept a compiled expression?

### ISSUE THREAD

**Issue by gwerbin** [pandas-dev/pandas#15446]
The current solution is to call `str.replace(<compiled_re>.pattern, flags=<compiled_re>.flags)` which is relatively ugly and verbose in my opnion.

Here's a contrived example of removing stopwords and normalizing whitespace afterwards:

```python
import pandas as pd
import re

some_names = pd.Series(["three weddings and a funeral", "the big lebowski", "florence and the machine"])

stopwords = ["the", "a", "and"]
stopwords_re = re.compile(r"(\s+)?\b({})\b(\s+)?".format("|".join(stopwords), re.IGNORECASE)
whitespace_re = re.compile(r"\s+")

# desired code:
# some_names.str.replace(stopwords_re, " ").str.strip().str.replace(whitespace_re, " ")

# actual code:
some_names.\
    str.replace(stopwords_re.pattern, " ", flags=stopwords_re.flags).\
    str.strip().str.replace(whitespace_re.pattern, " ", flags=whitespace_re.flags)
```

Why do I think this is better?

1. It's [nice](http://stackoverflow.com/a/453568/2954547) to have commonly used regular expressions compiled and to carry their flags around with them (and also allows the use of ["verbose" regular expressions](http://stackoverflow.com/a/29159014/2954547))
2. It's not that compiled regular expressions should quack like strings... it's that in this case we're making strings quack like compiled regular expressions, but at the same time not letting those compiled regular expressions quack their own quack.

Is there a good reason _not_ to implement this?

**Comment by TomAugspurger** [comment]
> Is there a good reason not to implement this?

I can't think of one. Feel free to give it a shot!

**Comment by jreback** [comment]
yup these reasonable (and straightforward) to do.

**Comment by rouzazari** [comment]
I just submitted a PR after looking into this (apologies in advance for the unexpected triple-commit message above). 

Very simply, I added a check to `pandas.core.strings.str_replace` to test whether `pat` is a compiled regular expression. 

A few questions that will help drive this to completion:

1. Testing whether a variable is a compiled regular expression could be moved to `pandas.types`. Is this preferred?
2. I'm not testing if `case == False` or `flags not None`, so a user may supply a compiled regular expression along with `case` and `flags` and receive output that does not incorporated `case` and `flags`. Is it preferred to handle these parameters as noop when provided with a compiled regular expression or to raise an error or warning? 
3. I can't find tests for `str_replace`. Where would I add test code for adding compiled regular expression handling?

Note:

- I'm using a protected variable `_re._pattern_type` to test if `pat` is a compiled regular expression. Alternatively this can be done by compiling a dummy regular expression just to get the type, but I'll wait until question 1 above is clear.
- Documentation in `str_replace` will depend on question 2 above.
